README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# SurvivalCurveExtraction

<!-- badges: start -->

[![Lifecycle:
experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://www.tidyverse.org/lifecycle/#experimental)
<!-- badges: end -->

This repository contains the codes necessary to produce an interactive
[shiny dash
board](https://pharmacoecon.shinyapps.io/SurvivalCurveExtraction/). The
original [Guyot et
al’s](https://bmcmedresmethodol.biomedcentral.com/articles/10.1186/1471-2288-12-9)
code is embedded in the dashboard. The original purpose of the document
was to serve as the analysis toolkit for generating survival
distribution parameters and fit statistics for health technology
assessment (HTA) or cost-effectiveness analysis of pharmaceutical
interventions.

## Purpose

1.  Allowing user to input datasets (as **.csv** files) and provide
    summary statistics of simulated Individual Patient Data (**IPD**)
    generated from survival plots of other studies.

2.  Users can specify which distributions to include (e.g. Weibull,
    Gompertz, and even Splines).

3.  Graphical display of the Kaplan-Meier plot and fitted parametric and
    semi-parametric distributions can also be extracted through the
    dashboard.

4.  Parameters of the parametric distributions and fit statistics can be
    obtained..

## Installation

You can install the released version of SurvivalCurveExtraction from
[CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("SurvivalCurveExtraction")
```

This package won’t likely be released so…

``` r
# devtools::install_github("ck2136/SurvivalCurveExtraction")
```

## Directions

There ar 3 main steps for extracting survival distribution parameters:

1.  Digitize the survival curve
2.  Create **.csv** files
3.  Input the **.csv** files to generate best fitting curve to generated
    survival data

### 1\. Digitizing the Survival Curve

1.  **Before** embarking on the use of the dashboard, the user is
    required to have the necessary data files extracted from previous
    studies as described in [Guyot et
    al](https://bmcmedresmethodol.biomedcentral.com/articles/10.1186/1471-2288-12-9).
    The first dataset should look like below:

<!-- end list -->

``` r
df1 <- read.csv(paste0(here(),"/data-raw/digitized_data.csv"), header = FALSE)
df1 %>%
  head %>%
  kable 
```

| V1 |        V2 |
| -: | --------: |
|  0 | 0.9954050 |
|  1 | 0.9988023 |
|  2 | 0.9946494 |
|  3 | 0.9729813 |
|  4 | 0.9683260 |
|  5 | 0.9635587 |

The first 6 rows of the **.csv** file is shown to illustrate that as
\(t\) goes from 0 to \(T\), the value of \(S(t)\) should be monotone
decreasing from \(1\) to \(0\).

This file can be extracted from the website
[WebPlotDigitizer](https://automeris.io/WebPlotDigitizer/). In fact
let’s walk through one example of a study using [maintenance Olaparib
in patients with newly diagnosed advanced Ovarian
cancer](https://www.nejm.org/doi/full/10.1056/NEJMoa1810858). The
survival curve that we’re going to try to extract is the Progression
Free Survival (PFS) curve for patients that were using Olaparib.

![](inst/app/www/moore_b_ola.png)

Once you download/extract the file from the website, you now want to
click the `Launch` button and load the image. You’ll see a screen like
below:

![](inst/app/www/CE_2.png)

Once the file is loaded and you select the Plot Type as `2D (X-Y) Plot`
click on the `Align Axes` button.

![](inst/app/www//wpd_axis.png)

The next step is to select the points `X1`,`X2`, `Y1`, and `Y2` and then
input the numeric values as a reference point.

![](inst/app/www/wpd_axes2.png)

`Proceed`\! **TIP:** Once you select each point, you can move the points
around (in order to make better calibrations) by using the keyboard
arrow keys.

Once you are done with selecting each of the points mentioned above,
click on the `Complete!` button. You’ll see a box appear where now you
enter in the values for each of the points `X1`,`X2`, `Y1`, and `Y2`.
For this exercise we want \(X1 = Y1 = 0\) and \(X2 = 57\) and
\(Y2 = 1\).

![](inst/app/www//wpd_axes3.png)

Now that we’ve set the reference points for our data points, now we need
to select the data points that is to be recorded. On the right had side
of the screen there are 4 buttons next to `Mask`. Usually for these
exercises I use a pen and draw over the line that I’ll be extracting as
such:

![](inst/app/www/wpd_extract.png)

Once you’ve drawn over the curve make sure the `Foreground Color` is the
appropriate color of the data points.

![](inst/app/www/wpd_foreground.png)

Now click on `Run` and you’ll see that red dots are generated\! We’re
almost done\! At this point we can play around with the `Algorithm` to
determine if we want to add more points or reduce the number of points.
Once you feel ready, we can go to the left hand side of the website and
click on `View Data`. Since we aren’t too concerned about going down to
\(>5\) digits, I usually format the code to have \(5\) fixed digits.
(Click on the `Format` button after selecting `Fixed` Digits). Then ou
can download the **.csv** file. I’ve named it `moore_b_ola.csv`.

We’ve successfully digitized the curve data hooray\!\!\!

### Create **.csv** Files

#### Data check for monotone decreasing

This step is very important as this is the part where we are going to
check the data for any inconsistencies. This data will later be used to
generate survival distribution parameters for simulations that will
impact the results\! Typically the results will be identifying whether
or not a certain treatment is more cost-effective compared to another.
If we do a poor job in this step, the results of this study will **NOT**
be appropriate and will lead to **BIASED** and **INCORRECT**
conclusions. Let’s do a good job in terms of data checking such that our
efforts here will yield appropriate conclusions.

The main point of this step is to identify if our data points are
non-increasing (ideally decreasing as per a usual survival curve). One
can visually check the data by plotting and also numerically check the
data. Below is a plot of the \(S(t)\) by \(t\).

``` r
plot(df1$V1, df1$V2, xlab = "t", ylab = "S(t)")
```

<img src="man/figures/README-unnamed-chunk-4-1.png" width="100%" />

It’s clear that the second data point has a greater \(S(t)\) value
compared to the first data point. This clearly needs to be adjusted. As
all survival curves usually have the first point as \(S(t) = 1\), we
should replace the value to be 1. The code `monotone_check()` takes the
original **.csv** file and converts it to a data file with \(S(t)\) as
being monotone decreasing. Below it is in action:

``` r
df1 %>%
  head
#>   V1        V2
#> 1  0 0.9954050
#> 2  1 0.9988023
#> 3  2 0.9946493
#> 4  3 0.9729813
#> 5  4 0.9683260
#> 6  5 0.9635587

df1edited <- monotone_check(df1)
par(mfrow=c(1,2))
plot(df1$V1, df1$V2, xlab = "t", ylab = "S(t)", main = "Original")
plot(df1edited$Time, df1edited$`S(t)`, xlab = "t", ylab = "S(t)", main = "Data Checked")
```

<img src="man/figures/README-unnamed-chunk-5-1.png" width="100%" />

For convenience, the `monotone_check()` function has been implemented in
the shinyapp so if the user doesn’t want to deal with it manually, the
application will deal with it foryou. (The user should still do a data
check to make sure that the \(S(t)\) are monotone decreasing).

Once the data has been checked for monotone decreasing, we can go ahead
and use the cleaned `df1edited` file as our first file indicated by
Guyot et al. The next **.csv** file needs to look like below:

``` r
df2 <- read.csv(paste0(here(),"/data-raw/digitized_nrisk.csv"), header = TRUE)
df2 %>%
  kable 
```

| Rownumber | Time | Lower.index | Upper.index | Number.at.risk |
| --------: | ---: | ----------: | ----------: | -------------: |
|         1 |    0 |           1 |           3 |            260 |
|         2 |    3 |           4 |           6 |            238 |
|         3 |    6 |           7 |           9 |            226 |
|         4 |    9 |          10 |          12 |            216 |
|         5 |   12 |          13 |          15 |            206 |
|         6 |   15 |          16 |          18 |            196 |
|         7 |   18 |          19 |          21 |            190 |
|         8 |   21 |          22 |          24 |            178 |
|         9 |   24 |          25 |          27 |            170 |
|        10 |   27 |          28 |          30 |            150 |
|        11 |   30 |          31 |          33 |            137 |
|        12 |   33 |          34 |          36 |            132 |
|        13 |   36 |          37 |          39 |            108 |
|        14 |   39 |          40 |          42 |             82 |
|        15 |   42 |          43 |          45 |             42 |
|        16 |   45 |          46 |          48 |             34 |
|        17 |   48 |          49 |          51 |              3 |
|        18 |   51 |          52 |          54 |              2 |
|        19 |   54 |          55 |          57 |              0 |

  - The columns need be as such:
    
      - Column 1: Row Number.
    
      - Column 2: Time in Months.
    
      - Column 3: Lower interval.
    
      - Column 4: Upper interval.
    
      - Column 5: Number at risk.

The values in each of the columns should reflect that of the values that
are from the number at risk information from the original survival
curve. At \(t = 0\), the number at risk for the Olaparib group is
\(260\). So our interval for number at risk during this period will be
\([0, 3)\). The Lower index value (**Column 3**) should correspond to
the index value in `df$edited$Index` which is 1. The Upper index value
(**Column 4**) should be just before 3, in our case therefore should be
\(t = 2.5\) which corresponds to \(\text{index} = 6\). Therefore,
following along this logic, the user needs to create a **.csv** file
that looks like `df2`.

Once the file has been created, we are ready to insert these **.csv**
files into the dashboard\!

### Input the **.csv** files to generate best fitting curve to generated survival data

There are several ways that one could start the dashboard application:

1.  Use
    [shinyapps.io](https://pharmacoecon.shinyapps.io/SurvivalCurveExtraction/)
2.  Clone/Fork the [github
    repo](https://github.com/ck2136/SurvivalCurveExtraction) and run the
    application using
    [Rstudio](https://rstudio.com/products/rstudio/download/#download)

Since it’s easier to use the web-application. I’m going to go with the
server. Click to access
[shinyapps.io](https://pharmacoecon.shinyapps.io/SurvivalCurveExtraction)

In the Introduction tab there is a guideline that everyone should be
able to follow.

![](inst/app/www/main.png)

The introduction is short and should be read everytime to reinforce the
requirements for the extraction of the survival parameter estimates, the
assumptions, and the workflow.

In the `Data Input and Summary` tab, we can enter our two `.csv` files
created. The [`/data-raw/`]() folder should have the files
`digitized_data.csv` and `digitized_nrisk.csv` for tutorial purposes.
Select the appropriate columns that represent \(S(t)\) and \(t\). The
numbes of events are \(102\) in the original study but you may play
around with that value to see how the parameters change.

*Note:* You will also notice that the `Results` tabe will show up once
you enter in both `.csv` files.

![](inst/app/www/datainput.png)

The `Results` tab will show

1.  Parameter estimates
2.  Fit Statistics
3.  Plots

With appropriate (monotone decreasing data that is sensible) data, the
desired outcomes should be achieved.

![](inst/app/www/parest.png)

## Contact

If there are any questions please do not hesitate to contact
[me](mailto:chong.kim@ucdenver.edu)

## Code of Conduct

Please note that the ‘SurvivalCurveExtraction’ project is released with
a [Contributor Code of Conduct](CODE_OF_CONDUCT.md). By contributing to
this project, you agree to abide by its terms.


NEWS.md
# SurvivalCurveExtraction 0.0.0.9001

* Added a Section in the `Parameter Estimation Tab` that incorporates variance-covariance matrix.

# SurvivalCurveExtraction 0.0.0.9000

* Added a `NEWS.md` file to track changes to the package.


R/app_ui.R
#' @import shiny
#' @import shinydashboard
#' @import flexsurv
#' @import DT
#' @import ggplot2
#' @import ggfortify
#' @import plotly
#' @import dplyr
#' @import MASS
#' @import splines
#' @import survival
#' @import formattable
#' @import shinyAce
#' @import survminer
app_ui <- function() {
  tagList(
    # Leave this function for adding external resources
    golem_add_external_resources(),
    # List the first level UI elements here 
    
    dashboardPage(
      
      # Header
      dashboardHeader(title = "Survival Curve Extraction Dashboard"),
      
      
      # Sidebar Menus
      dashboardSidebar(
        conditionalPanel("output.data1exists == true && output.data2exists == true",
                         sidebarMenu(id = "sidbarmenu",
                                     menuItem("Introduction", tabName = "dashboard", icon = icon("dashboard")),
                                     # menuItem("Extraction and Error Check", tabName = "extractData", icon = icon("th")),
                                     menuItem("Data Input and Summary", tabName = "data1", icon = icon("th"),
                                              # collapsesible = 
                                              #   menuSubItem('random1', tabName = 'random1'),
                                              menuSubItem('Data Input', tabName = 'data'),
                                              menuSubItem('Data Check', tabName = 'datacheck'),
                                              menuSubItem('KM estimate', tabName = 'datasum')
                                     ),
                                     menuItem('Results', tabName = 'results', icon = icon('th'), 
                                              # collapsible = 
                                              #   menuSubItem('random', tabName = 'random'),
                                              menuSubItem('Parameter Estimates', tabName = 'regresult'),
                                              menuSubItem('Fit Statistics', tabName = 'diag'),
                                              menuSubItem('Plots', tabName = 'plot')
                                     ),
                                     menuItem("Software Info", tabName = "softInfo", icon = icon("th"))
                                     
                         )
                           
        ),
        conditionalPanel("output.data1exists != true || output.data2exists != true",
                         sidebarMenu(id = "sidbarmenu",
                                     menuItem("Introduction", tabName = "dashboard", icon = icon("dashboard")),
                                     # menuItem("Extraction and Error Check", tabName = "extractData", icon = icon("th")),
                                     menuItem("Data Input and Summary", tabName = "data1", icon = icon("th"),
                                              collapsesible = 
                                                menuSubItem('random1', tabName = 'random1'),
                                              menuSubItem('Data Input', tabName = 'data'),
                                              menuSubItem('Data Check', tabName = 'datacheck'),
                                              menuSubItem('KM estimate', tabName = 'datasum')
                                     ),
                                     menuItem("Software Info", tabName = "softInfo", icon = icon("th"))
                                     
                         )
                           
        )
        
      ),
      
      # Body of the Dashboard
      dashboardBody(
        tabItems(
          
          # First tab content DashBoard
          tabItem(tabName = "dashboard",
                  
                  h2("Introduction"),
                  fluidRow(
                    box(
                      h4('This is a dashboard for generating survival curve distribution parameters, fit statistic, and figures that are useful and/or necessary for economic simulations. Example of health technology assessment using the R package ', a("heemod", href="https://cran.r-project.org/web/packages/heemod/vignettes/a_introduction.html", style = "color:red"), " is ", a("here", href="https://cran.r-project.org/web/packages/heemod/vignettes/j_survival.html", style = "color:red")),
                      
                      h4("There are several important key points to remember:", style = "color:red"),
                      
                      h4(strong("1."),"User needs to have 2 .csv files to input which is described in ", a("Guyot et al. 2012", href = "https://bmcmedresmethodol.biomedcentral.com/articles/10.1186/1471-2288-12-9", style = "color:red; font-style: italic")),
                      
                      h4("If ",strong("1")," is not satisfied: ","User needs to extract the survival curve data points by using a figure digitizer such as ", a("Webplotdigitizer", href="https://apps.automeris.io/wpd/", style = "color:red"), " and extract the curve data and relevant risk and censoring information as described in ", a("Guyot et al. 2012", href = "https://bmcmedresmethodol.biomedcentral.com/articles/10.1186/1471-2288-12-9", style = "color:red; font-style: italic")),
                      
                      h4(strong("2."),"Before trying to determine a survival distribution that best fits the data, user needs to check that the extracted curve is as accurate as possible. Human error is highly possible to be introduced during the process of survival curve extraction thus this step is crucial in terms of preserving the integrity of the original study's survival information"),
                      
                      h4(strong("3."),"The outputs that are generated in the 'Result' tab (Bottom tab on the left) requires the user to have a basic understanding of the analysis of time to event data. An understanding of parameters for survival distributions and fit statistics will be very helpful. Optionally, understanding of spline functions in fitting a survival distribution to the data may be beneficial"),
                      h4(strong("Everyone should check for data entry errors!"), " regardless of how well the curve was digitized/constructed, anyone who has ever dealt with manual extraction should check the curve visually as well as numerically in terms of the S(t) for each values of t. Please check the 'Extraction and Error Check' tab on the left side an go through the exercise"),
                      
                      h4("A short tutorial of the process of extracting survival curve parameters is illustrated on", a("github", href="https://github.com/ck2136/SurvivalCurveExtraction", style = "color:red")," for those that may need a vignette. If you have any questions regarding how to go about each step please consult", a("Guyot et al. 2012", href = "https://bmcmedresmethodol.biomedcentral.com/articles/10.1186/1471-2288-12-9", style = "color:red; font-style: italic"), " and if still difficult to proceed, e-mail: ", a("chong.kim@ucdenver.edu", href = "mailto:chong.kim@ucdenver.edu"))
                      
                      ,width = 12)
                  )
                  
          ),
          
          # Second Tab content: Information for gathering csv files
          # tabItem(tabName = "extractData",
          #         
          #         fluidRow(
          #           box(
          #             h2("Extraction and Error Checking"),
          #             
          #             h4('If the required survival curve and number at risk csv files are not extracted beforehand, please follow the', a("Tutorial", href="https://github.com/ck2136/SurvivalCurveExtraction/blob/master/SurvivalExtractionforCEA.ipynb"),' so that the user can extract the curve according to Guyot et al. then be able to generate survival distribution parameter estimates')
          #             ,width = 12
          #           )
          #         ),
          #         
          #         fluidRow(
          #           box(
          #             h4(strong("Preview of the tutorial")),
          #             
          #             img(src='www/step1.gif', align = 'center')
          #             , width = 12 , align = 'center'
          #           )
          #         ),
          #         
          #         fluidRow(
          #           column(8,
          #                  box(
          #                    h4(strong("Preview of Number 2")),
          #                    
          #                    img(src='www/step2.gif', align = 'center'),
          #                    
          #                    h4("Currently the most important aspect with regard to checking whether or not the survival curve is appropriate is to determine if it is monotone decreasing (i.e. not increasing). Due to approximation errors in the webplotdigitizer, there may be points that have higher S(t) than the previous time point. The tutorial in the above Tutorial link has python code to remedy non-monotonic S(t) but there are also ways to do this in excel.")
          #                    , width = 8, align = 'center'
          #                  ), offset = 3
          #           )
          #           
          #         )
          # ),
          
          # Second Tab content: Data and SUmmary
          tabItem(tabName = "data",
                  
                  box(
                    
                    h2("Digitized survival curve input"),
                    box(
                      fileInput('datafile1', 'Please input the digitized survival curve csv file',
                                accept=c('text/csv', 'text/comma-separated-values,text/plain')) 
                      ,
                      width = 6
                    ),
                    box(
                      uiOutput("Variables1"),
                      uiOutput("Variables2"),
                      width = 6
                    ),
                    width = 12
                  )
                  ,
                  
                  box(
                    
                    h2("Number at risk input"),
                    box(
                      fileInput('datafile2', 'Please input the number at risk csv file',
                                accept=c('text/csv', 'text/comma-separated-values,text/plain')),
                      width = 6
                    ),
                    box(
                      numericInput('totevent',"Total number of events",value =  NA, min = 0),
                      numericInput('arm.id',"Arm id: control = 0; treatment = 1", 1, min = 0, max = 1),
                      width = 6
                    ),
                    width = 12
                  )
                  
          ),
          
          tabItem(
            tabName = "datacheck",
            fluidRow(
              width = 10,
              box(
                plotOutput("orgplot", width = 700), width = 10
              )
            )
          ),
          
          # Data SUmmary tab content
          
          tabItem(tabName = 'datasum',
                  
                  fluidRow(
                    box(
                      # h3("Individual Patient Data (IPD) View"),
                      # dataTableOutput("IPDtable"),
                      
                      h3("KM Estimate"),
                      verbatimTextOutput("KMsum")
                      , width = 12
                    )
                  )
                  
          ),
          
          # Plots tab: Plot submenu of result
          tabItem(tabName = "plot",
                  
                  fluidRow(
                    box(
                      h3("Plotting Options"),
                      box(
                        radioButtons('confint','Confidence Intervals?',
                                     choices =  c(
                                       "Yes" = 'yes',
                                       "No" = 'no'), selected = 'no'),
                        radioButtons('wpdcurve','Add Digitized Curve?',
                                     choices = c(
                                       "Yes" = 'yes',
                                       "No" = 'no'), selected = 'no'
                        ),
                        width = 4
                      ),
                      box(
                        # Copy the line below to make a text input box
                        textInput("titletxt", label = "Title", value = "Simulated survival distributions"),
                        textInput("xlab", label = "X Label", value = "Time in Months"),
                        width = 4
                      ),
                      box(
                        # Copy the line below to make a text input box
                        textInput("ylab", label = "Y Label", value = "S(t)"),
                        radioButtons('splineyn','Splines?',
                                   choices =  c(
                                     "Yes" = 'yes',
                                     "No" = 'no'), selected = 'no'),
                        conditionalPanel("input.splineyn == 'yes'",
                                   sliderInput('numspline', "Number of splines:",
                                               min = 1, max = 100, value = 1, step = 1)
                  ),
                        width = 4
                      ),
                      width = 12)
                  ),
                  fluidRow(
                    width = 10,
                    box(
                      plotOutput("survplot", width = 700), width = 10
                    )
                  )
          ),
          
          
          # Parameter Estimates tab content
          tabItem(tabName = "regresult",
                  
                  h2("Parameter Estimates"),
                  
                  fluidRow(
                    box(
                      h3("Survival Regression Parameters"),
                      fluidRow( tableOutput("parest"))
                      ,width = 12
                    )
                    
                    
                  ),
                  fluidRow(
                    box(
                      h3("Covariance Matrix"),
                      fluidRow( tableOutput("vcovmat")),
                      fluidRow( downloadButton('downloadvcov', "Download"))
                      ,width = 12
                    )
                    
                    
                  )
                  
          ),
          
          # Sixth tab: Fit statistics/Diagnostic
          tabItem(tabName = "diag",
                  h2("Fit Statistics"),
                  
                  fluidRow(
                    box(
                      h3("Fit Statistics"),
                      fluidRow( tableOutput("fitstat"))
                      ,width = 12
                    )
                  )
                  
                  
          ),
          tabItem(tabName = "softInfo",
                  h2("Current R Session Information"),
                  
                  fluidRow(
                    box(
                      verbatimTextOutput("info")
                      ,width = 12
                    )
                  )
                  
                  
          )
          
          
          
        )
      )
    )
  )
  
}

#' @import shiny
golem_add_external_resources <- function(){
  
  addResourcePath(
    'www', system.file('app/www', package = 'SurvivalCurveExtraction')
  )
 
  tags$head(
    golem::activate_js(),
    golem::favicon()
    # Add here all the external resources
    # If you have a custom.css in the inst/app/www
    # Or for example, you can add shinyalert::useShinyalert() here
    #tags$link(rel="stylesheet", type="text/css", href="www/custom.css")
  )
}
